### ISSUE TITLE
Print warnings with errors?

### ISSUE THREAD

**Issue by goldfirere** [ghc-proposals/ghc-proposals#411]
@adinapoli and I are cleaning out some accumulated dust in the error-handling infrastructure, en route to implementing structured error messages (#306).

In my work on this this afternoon, I came across some interesting behavior, and I don't know whether it's wrong, and if it is, in which direction to fix it.

First off, consider

```hs
foo x = not ()
```

GHC reports that `()` does not match `Bool`. It does *not* report anything about unused variable `x`, even with warnings enabled. This is because reporting of warnings is suppressed when there are errors about.

However, let's examine

```hs
foo x = \x -> x
```

GHC reports three warnings: one because I've made a top-level definition without a type signature, one because I didn't use the first `x`, and one because my second `x` shadows the first. All good.

Now I recompile with `-Werror=name-shadowing`. Interestingly, I now see two warnings and an error -- something I had not thought possible in GHC.

So:

1. Should GHC suppress warnings when there are errors?
2. If answering "yes" to (1), then should warnings-promoted-into-errors with `-Werror` also be suppressed? ("Yes" means that warnings-promoted-into-errors are treated more like warnings here.)
2. If answering "yes" to (1), then should warnings-promoted-into-errors with `-Werror` also suppress? ("No" means that warnings-promoted-into-errors are treated more like warnings here.)

Right now, GHC's implementation answers "yes", "yes", and "no". To me, this feels like an odd design choice. If the user writes `-Werror`, then I would think that warnings-promoted-into-errors get treated just like errors. But maybe this is too rigid, and the current design is more useful in practice.

My choice would be to answer "no" to (1): just print out all that we've got. If you don't want the errors, use `-w`.

What do you think?

(Note: this isn't really a GHC proposal, because it's just fiddling with diagnostics, which are generally considered just beyond the scope of proposals.)

**Comment by ndmitchell** [comment]
Note that for an IDE, the best answer by a significant margin is 1. It's easy for the IDE to put warnings at the top, but having access to all of them often gives you a clue about what the error means. E.g. it's not uncommon to see the error `no function named foo123457` alongside the warning `unused function foo123456`.

**Comment by goldfirere** [comment]
By "the answer is ... 1", I assume you mean you want the answer to (1) to be "no". This seems pretty clear from your commentary, but I just want to make sure.

**Comment by ndmitchell** [comment]
To confirm, yes, GHC should not hide warnings just because there are errors.

**Comment by tomjaguarpaw** [comment]
I agree with Neil when it comes to IDEs. I'm frequently disappointed that dante can only highlight the errors or warnings that GHC thinks is the most important, rather than the errors or warnings that I actually want to fix first. When it comes to command line use I don't know.

**Comment by edsko** [comment]
Strong vote for "yes" on question 1 (though I understand the IDE needs are different; perhaps this itself should be a flag).

PureScript shows warnings even when there are errors and I find it hugely distracting. "Yes, I _know_ that variable is unused, I'm still working on this code!". Sure, with the right tooling that could be made much better, but I find ghc's default (showing warnings only once the errors are resolved) very sensible. Yes, we could disable warnings, but now I need to toggle this all the time; "disable warnings.. fix the error.. enable them again.. repeat".

The use of `-Werror` is primarily to make sure that the code is `-Wall` clean, but again, that's not something you really want to think about when the code is still actually broken, so suppressing these warnings-promoted-to-errors while there are genuine erorrs I think is reasonable. This one I feel slightly less strongly about though because toggling `-Werror` from time to time ("is my code base still -Wall clean?" is doable, in principle.

I don't understand question 2 versus question 3.

**Comment by adinapoli** [comment]
Thanks folks, that's very valuable input. I will play the devil's advocate here (for the good or the bad, I guess 😉 ): for IDEs, this will make perfect sense, and indeed the work we are doing to implement #306 has as one of the main goals exactly making IDEs' authors (and users) happier and their lives easier. 

However, I wonder about end users that are not using an IDE (for any reason): maybe HLS/dante/intero/.. never worked for them, they can't switch for various reasons or they are simply happy with their current workflow which might involve nothing more than calling `cabal/stack` in a separate terminal, checking the diagnostics, fixing them, rinse and repeat.

If I would fall into this category I guess I would be a bit upset to see that now GHC buries my errors into a few dozens of irrelevant warnings. There are circumstances where you want to have something compile quickly and easily, and it's not very helpful if GHC is now telling me that `Control.Concurrent.MVar` and another dozen of modules I happened to have imported just to try things out are redundant _while_ I am trying to fix a tricky type error. Maybe the example is a bit artificial, but I guess the general idea remains. This is also pretty much in line with what @edsko is saying.

I am now thinking that indeed this should be configurable by a flag: if we had something like `-freport-all-diagnostics` or similar, which is turned off by default, this will bring to the table several benefits:

* IDEs can still trivially invoke GHC with this flag, and get all the different diagnostics;
* End users won't be affected by this change;
* Existing tests won't be affected by this change, but  we can easily write _new_ tests to check for the new behaviour;
* It should be easy to implement: we currently do this "keep/discard" dance IIRC in the Driver, so it should be a matter of checking for that flag only at that point.

**Comment by tdammers** [comment]
The crux, as I understand it:

- We have both IDE's and humans as consumers of the "error output API".
- The current "API" was designed with humans in mind, and it does a good job.
- The current "API" does not do a good job at serving IDE's; that is part of why we're doing this whole thing in the first place.
- Ultimately, the goal is to have two separate ways of getting error information out of GHC: one for humans, similar to the status quo, and one for IDE's, structured, rigid, exhaustive, and ideally following a versioned specification.

Adding a flag that, while not changing the *format*, changes the *filtering* of error information, would be a great first step: we need this mechanism anyway (so the effort isn't lost), and it would provide a payoff for the "IDE" side already without sacrificing anything on the "humans" side.

As for those questions:

1. For the "humans" case: a resounding **"yes".** Error output space is a scarce resource, and when errors are present, "errors are more important/interesting than warnings" is about as good a heuristic as we can get. For the "IDE" case, the answer is simply "report everything, always".
2. Also a resounding **"yes"**. The reason we use `-Werror` is not to change which warnings we get to see, hence the showing/hiding behavior should not change. The reason we use it is because we want the build to fail in the presence of warnings. But the "errors are more important than warnings" heuristic still applies. (In the "IDE" case, this does not apply: here, it's still "report everything, always").
3. From the previous, it follows that the answer should be **"no"**. Just like warnings in `-Werror` mode are still suppressed by "proper" errors, and thus still treated as warnings for this part of the pipeline, they should also be treated as "non-masking" themselves. (And of course in the "IDE" case, again, it's "report everything, always").

In a nutshell: for humans, don't change a thing; for machines, report everything, always.

**Comment by maralorn** [comment]
IDEs (hls) can and do work around this today (I think by using defer-type-errors?). So I think the default should be convenient for terminal users.

What is btw. my biggest problem with error reporting is, that any renamer error makes all typecheck errors disappear. (I think?) That’s quite unintuitive for people not familiar with the compilation process and can give you the wrong sense of "hah, yeah, that function seems to typecheck". Same goes for parsing errors but that’s probably obvious enough …

**Comment by JakobBruenker** [comment]
On a related note, I opened [this issue](https://gitlab.haskell.org/ghc/ghc/-/issues/20573) recently about wanting `-Wtabs` to be displayed even when there are errors

**Comment by goldfirere** [comment]
Thanks for the commentary here. I don't think keeping this Issue open helps us, so I'm closing.
